### ISSUE TITLE
Benchmarking

### ISSUE THREAD

**Issue by brainstorm** [umccr/htsget-rs#46]
As mentioned on the original GSoC proposals, this project is not complete until we validate that is well performant :)

For this, I've been cooking this repo: https://github.com/umccr/aws-benchmarks

As an example, using AWS S3 client libraries as subjects to test both CPU and I/O throughput on... but the idea is to apply those benches over here.

This task depends on #45 being finished.

**Comment by brainstorm** [comment]
Other than measuring throughput on the different htsget-rs layers, we should keep a close eye on this common htsget problem and metric: https://github.com/googlegenomics/htsget/issues/7

**Comment by brainstorm** [comment]
I’d argue that the best course of action is to create a benches subdirectory for each htsget-rs crate instead of creating a dedicated “htsget-benchmarks” crate within the workspace (as the `aws-benchmarks` crate).

I’d love the benchmarks for the server to be as end-to-end to measure real world throughput.

**Comment by victorskl** [comment]
Yah, benchmarks for the server sound good. On top of my head

- pick and stage some (large enough) public dataset (a BAM, a VCF)
- config server to serve from this dataset
- prepare few target slices -- i.e. pick different variant regions / alignment references -- may be 10 to 20s in a array
- baseline
  - try sequential calls to (hts)get each target slices; one by one
  - try simultaneous calls to get all target slices at once
- varying bench settings 
  - try apply any compiler optimisation, if any
  - try apply platform specific optimisation flag, if any
  - etc...
 
And by then by doing so, you could come up with more methodology around it, at that point... 

This need to do for both referencing impl and Rust impl here. You can start with local for this arrangement; and then by reaching at the point; it will be clearer what come next, I reckon.

Please also to make sure implementation "correctness", first i.e. server responses are correct according to the Spec and/or same with the referencing impl. If anything differ, need to note it.

**Comment by brainstorm** [comment]
Thanks a ton Victor for the thoughts and guidelines! There's another bit I just recalled and that is the amount of MB's this implementation returns for a given query.

I'm thinking and referring to this concerning issue with htsget: https://github.com/googlegenomics/htsget/issues/7

**Comment by CastilloDel** [comment]
Well, I still need to tinker a bit to see which problems I will find, but Victor's guidelines seem to make a lot of sense for now.

**Comment by CastilloDel** [comment]
I'm having problems with the reference implementation 😭 After following the instruction in the [README](https://github.com/googlegenomics/htsget#building-the-server) for building and running the server, I'm getting this error:
```
samtools flagstat http://localhost:8081/reads/public-bucket/test.bam
[E::hts_open_format] Failed to open file "http://localhost:8081/reads/public-bucket/test.bam" : Permission denied
samtools flagstat: Cannot open input file "http://localhost:8081/reads/public-bucket/test.bam": Permission denied
```
I didn't find anything like this in the issues, should I add a new one? The project seems somewhat abandoned, with the most recent issue/PR being from 2019 😥 
Also, it [can't be used as a local server](https://github.com/googlegenomics/htsget/issues/50) so comparing the actual state of our implementation with it might be pointless (?).

**Comment by brainstorm** [comment]
Oh, please don't try to deploy the googlegenomics htsget, it's indeed quite Google-tailored. I pointed at that particular issue because I wanted to know how many extra reads *the GA4GH reference implementation* returned, not Google's one, sorry for not being clear there :-S

**Comment by CastilloDel** [comment]
Seems like I got some things mixed up. Where can I find the reference implementation?

**Comment by brainstorm** [comment]
Sorry, when I say reference implementation, I usually refer to GA4GH's one here:

https://github.com/ga4gh/htsget-refserver

... perhaps sooner our later ours (htsget-rs) will be the GA4GH "reference implementation", we'll see after the benchmarking ;)

**Comment by CastilloDel** [comment]
That would be awesome! 💯

**Comment by CastilloDel** [comment]
I think I have hit a blocker :(. I've been using reqwest to perform the client requests in the benchmarks. Until now, I had only tried to get the initial response from the servers, but now I tried to also download the files with the urls the servers provided and I found a problem with our implementation. The urls it gives use the "file://" scheme, which isn't allowed by reqwest. This was something that already needed to be changed, because the htsget spec doesn't allow it, but we might have to deal with it sooner than expected.
